Users/stevenh/node utils

stevenhorsman opened the pull request:
#### PR type
- [X] Other... Add more component and extender properly tested SPI scripts

#### Relevant issues
Part of #1336 

#### Changes proposed in this PR
- Add node utils function and refactor utils error reporting

stevenhorsman commented:
@jackjia-ibm - if you get a chance to look this is the next set of utils changes with some of the refactoring from the previous PR done, but probably some more stuff that you can think of ways to improve

jackjia-ibm commented on bin/utils/file-utils.sh:
```diff
@@ -10,7 +10,27 @@
 ################################################################################
 
-#TODO LATER - provide flag that toggles all functions to error if they exit non-zero?
+# TODO LATER - anyway to do this better?
```
Not sure a better way, one thing we may be able to try to consolidate the variable names `INSTALL_DIR`, `ZOWE_ROOT_DIR`, and `ROOT_DIR`. But seems not the scope of this PR.

stevenhorsman replied:
Agreed - created https://github.com/zowe/zowe-install-packaging/issues/1373 to cover

stevenhorsman commented:
Passing convenience tests: https://wash.zowe.org:8443/job/zowe-install-test/view/change-requests/job/PR-1369/6/ and SMP/E PTF test: https://wash.zowe.org:8443/job/zowe-install-test/view/change-requests/job/PR-1369/7/

jackjia-ibm commented on bin/utils/node-utils.sh:
```diff
@@ -0,0 +1,144 @@
+    return 1
+  fi
+
+  node_ok=`${NODE_HOME}/bin/node -e "console.log('ok')" 2>&1`
```
This line has a chance to hang in a rare case. So maybe we echo something before this line to indicate we are validating if node bin is functional. So when we checking the log, we are sure the flow hangs on this line.

Also seems better this section moved below `node -v` check.

stevenhorsman replied:
I've added the logs and moved this check below the version, which required some refactoring in order to still test

stevenhorsman commented on bin/utils/node-utils.sh:
```diff
@@ -0,0 +1,144 @@
+
+# TODO - how to test well given interaction and guess?
+# Interactive function that checks if the current NODE_HOME is valid and if not requests a user enters the node home path via command line
+prompt_node_home() {
```
I guess there are a couple of problems:
1. These scripts are called a fair way down the stack, so passing a flag through all is not great. We could make a opt on zowe-configure-instance.sh/zowe-setup-certificates.sh to make it non interactive and export a shell variable based on that? I'm not sure if `-i` would be confusing for non-interactive though!
2. What behaviour do we want for non-interactive? We obviously don't want to exit the script, so just print an error and continue?

jackjia-ibm replied:
`NODE_HOME` is mandatory in this case and for future steps to start Zowe, so I think if it's non-interactive mode, it's safe we prompt the error and exit the script.

stevenhorsman replied:
My concern is that not all of our scripts are idempotent, so I don't know if we will hit issues when the user fixes the issue then re-runs. Do you mind if I create a new issue for this rather than have to solve it right now?

jackjia-ibm replied:
Yes that sounds good to me. Thanks!

stevenhorsman replied:
Captured in https://github.com/zowe/zowe-install-packaging/issues/1374

stevenhorsman commented on bin/utils/common.sh:
```diff
@@ -0,0 +1,42 @@
+
+#TODO LATER - do we want to provide a ENV_VAR flag that toggles whether errors are printed or not?
+
+print_error_message() {
```
Add print_message

jackjia-ibm commented on bin/utils/node-utils.sh:
```diff
@@ -0,0 +1,144 @@
+    return 1
+  fi
+
+  node_version=`${NODE_HOME}/bin/node --version`
```
`node -v` has chance to display an error rather than exact version number. When this happens, the flow will run in a dead loop kept complaining the version is not supported. We should detect and notify the user if this command hits an error and let the user to exit the flow to fix it offline.

FYI, this also happens on `java -version`.

stevenhorsman replied:
Added a new check and test for if we get an error when calling version

stevenhorsman commented:
Tests: https://wash.zowe.org:8443/job/zowe-install-test/view/change-requests/job/PR-1369/10/ & https://wash.zowe.org:8443/job/zowe-install-test/view/change-requests/job/PR-1369/9

jackjia-ibm commented on bin/utils/node-utils.sh:
```diff
@@ -0,0 +1,159 @@
+    fi
+  done
+  export NODE_HOME=$NODE_HOME
+  echo "  NODE_HOME variable value="${NODE_HOME} >> $LOG_FILE
```
Sorry didn't catch this early, this line should use `log_message` function.

stevenhorsman replied:
Fixed

jackjia-ibm commented on bin/utils/node-utils.sh:
```diff
@@ -0,0 +1,159 @@
+ensure_node_is_on_path() {
+  if [[ ":$PATH:" != *":$NODE_HOME/bin:"* ]]
+  then
+    echo "Appending NODE_HOME/bin to the PATH..."
```
Same reason, this line should use `print_message`. This function will be executed in `run-zowe.sh`.

stevenhorsman replied:
Fixed

stevenhorsman commented:
New tests:
- https://wash.zowe.org:8443/job/zowe-install-test/view/change-requests/job/PR-1369/11/
-https://wash.zowe.org:8443/view/Zowe/job/zowe-install-test/view/change-requests/job/PR-1369/15/
Fixed issues and refactored pre-reqs to remove duplicated code:
```Checking Node version...
Node version v8.17.0 is supported
Validating if node bin is functional...
Node bin is functional
Script zowe-check-prereqs.sh ended
```
